### PR TITLE
corrects loss function for Self-play Preference Optimization hard label version

### PR DESCRIPTION
**Pull request by angelahzyuan**
corrects implementation mentioned in https://github.com/huggingface/trl/pull/1612. arxiv: https://arxiv.org/abs/2405.00675

### CODE REVIEW

**Comment by angelahzyuan** [comment]
@winglian

**Review by kashif — APPROVED** [review]

**Comment by kashif** [comment]
@angelahzyuan thank you for the fix! Can you kindly run the `pre-commit run --all-files` command in the root of `trl` to fix the formatting etc.

**Comment by kashif** [comment]
@angelahzyuan perhaps move the comment inside the `elif` of the loss, currently its outside, to avoid confusion

**Comment by angelahzyuan** [comment]
@kashif done

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -1028,14 +1028,15 @@ def dpo_loss(
             losses = -F.logsigmoid((self.beta * chosen_logratios) - delta) - F.logsigmoid(
                 -(self.beta * rejected_logratios - delta)
             )
-        elif self.loss_type == "sppo":
-            a = self.beta * (policy_chosen_logps - reference_chosen_logps)
-            b = self.beta * (policy_rejected_logps - reference_rejected_logps)
+        elif self.loss_type == "sppo_hard":
+        # In the paper (https://arxiv.org/pdf/2405.00675), SPPO employs a soft probability approach, estimated using the PairRM score. The probability calculation is conducted outside of the trainer class. The version described here is the hard probability version, where P in Equation (4.7) of Algorithm 1 is set to 1 for the winner and 0 for the loser.
```
Comment (kashif) [inline trl/trainer/dpo_trainer.py:1032]:
```suggestion
            # In the paper (https://arxiv.org/pdf/2405.00675), SPPO employs a soft probability approach, estimated using the PairRM score. The probability calculation is conducted outside of the trainer class. The version described here is the hard probability version, where P in Equation (4.7) of Algorithm 1 is set to 1 for the winner and 0 for the loser.
```

**Comment by cx9208** [comment]
i got loss=50.0 when set beta=0.1,  what is beta mean in sppo_hard and how to set correct beta?thx

**Pull request merged**